db-migration/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+use ckb_db::RocksDB;
```

Comment:
ℹ️ Moved from [db/src/migration.rs](https://github.com/nervosnetwork/ckb/pull/2119/files#diff-3cb74f314b728df4c86dba728183a6fd)

Reply:
This is a diff of the two files

``` diff
diff --git a/db/src/migration.rs b/db-migration/src/lib.rs
index 6ef8f7b0..1e91a307 100644
--- a/db/src/migration.rs
+++ b/db-migration/src/lib.rs
@@ -1,9 +1,12 @@
-use crate::{
-    db::{RocksDB, VERSION_KEY},
-    internal_error, Result,
-};
+use ckb_db::RocksDB;
+use ckb_error::{Error, InternalErrorKind};
 use ckb_logger::info;
-use rocksdb::ops::{Get, Put};
+
+pub const VERSION_KEY: &[u8] = b"db-version";
+
+fn internal_error(reason: String) -> Error {
+    InternalErrorKind::Database.reason(reason).into()
+}
 
 #[derive(Default)]
 pub struct Migrations {
@@ -11,14 +14,17 @@ pub struct Migrations {
 }
 
 impl Migrations {
+    pub fn new() -> Self {
+        Migrations { migrations: vec![] }
+    }
+
     pub fn add_migration(&mut self, migration: Box<dyn Migration>) {
         self.migrations.push(migration);
     }
 
-    pub fn migrate(&self, db: &RocksDB) -> Result<()> {
+    pub fn migrate(&self, mut db: RocksDB) -> Result<RocksDB, Error> {
         let db_version = db
-            .inner
-            .get(VERSION_KEY)
+            .get_pinned_defalt(VERSION_KEY)
             .map_err(|err| {
                 internal_error(format!("failed to get the version of database: {}", err))
             })?
@@ -27,33 +33,31 @@ impl Migrations {
             });
 
         match db_version {
-            Some(v) => self
-                .migrations
-                .iter()
-                .filter(|m| m.version() > v.as_str())
-                .try_for_each(|m| {
-                    info!("Migrating database version to {}", m.version());
-                    m.migrate(db)?;
-                    db.inner.put(VERSION_KEY, m.version()).map_err(|err| {
+            Some(v) => {
+                for m in self.migrations.iter().filter(|m| m.version() > v.as_str()) {
+                    db = m.migrate(db)?;
+                    db.put(VERSION_KEY, m.version()).map_err(|err| {
                         internal_error(format!("failed to migrate the database: {}", err))
-                    })
-                }),
+                    })?;
+                }
+                Ok(db)
+            }
             None => {
                 if let Some(m) = self.migrations.last() {
                     info!("Init database version {}", m.version());
-                    db.inner.put(VERSION_KEY, m.version()).map_err(|err| {
+                    db.put(VERSION_KEY, m.version()).map_err(|err| {
                         internal_error(format!("failed to migrate the database: {}", err))
-                    })
-                } else {
-                    Ok(())
+                    })?;
                 }
+                Ok(db)
             }
         }
     }
 }
 
 pub trait Migration {
-    fn migrate(&self, db: &RocksDB) -> Result<()>;
+    fn migrate(&self, _db: RocksDB) -> Result<RocksDB, Error>;
+
     /// returns migration version, use `yyyymmddhhmmss` timestamp format
     fn version(&self) -> &str;
 }
@@ -71,11 +75,116 @@ impl DefaultMigration {
 }
 
 impl Migration for DefaultMigration {
-    fn migrate(&self, _db: &RocksDB) -> Result<()> {
-        Ok(())
+    fn migrate(&self, db: RocksDB) -> Result<RocksDB, Error> {
+        Ok(db)
     }
 
     fn version(&self) -> &str {
         &self.version
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ckb_app_config::DBConfig;
+
+    #[test]
+    fn test_default_migration() {
+        let tmp_dir = tempfile::Builder::new()
+            .prefix("test_default_migration")
+            .tempdir()
+            .unwrap();
+        let config = DBConfig {
+            path: tmp_dir.as_ref().to_path_buf(),
+            ..Default::default()
+        };
+        {
+            let mut migrations = Migrations::default();
+            migrations.add_migration(Box::new(DefaultMigration::new("20191116225943")));
+            let r = migrations.migrate(RocksDB::open(&config, 1)).unwrap();
+            assert_eq!(
+                b"20191116225943".to_vec(),
+                r.get_pinned_defalt(VERSION_KEY).unwrap().unwrap().to_vec()
+            );
+        }
+        {
+            let mut migrations = Migrations::default();
+            migrations.add_migration(Box::new(DefaultMigration::new("20191116225943")));
+            migrations.add_migration(Box::new(DefaultMigration::new("20191127101121")));
+            let r = migrations.migrate(RocksDB::open(&config, 1)).unwrap();
+            assert_eq!(
+                b"20191127101121".to_vec(),
+                r.get_pinned_defalt(VERSION_KEY).unwrap().unwrap().to_vec()
+            );
+        }
+    }
+
+    #[test]
+    fn test_customized_migration() {
+        struct CustomizedMigration;
+        const COLUMN: &str = "0";
+        const VERSION: &str = "20191127101121";
+
+        impl Migration for CustomizedMigration {
+            fn migrate(&self, db: RocksDB) -> Result<RocksDB, Error> {
+                let txn = db.transaction();
+                // append 1u8 to each value of column `0`
+                let migration = |key: &[u8], value: &[u8]| -> Result<(), Error> {
+                    let mut new_value = value.to_vec();
+                    new_value.push(1);
+                    txn.put(COLUMN, key, &new_value)?;
+                    Ok(())
+                };
+                db.traverse(COLUMN, migration)?;
+                txn.commit()?;
+                Ok(db)
+            }
+
+            fn version(&self) -> &str {
+                VERSION
+            }
+        }
+
+        let tmp_dir = tempfile::Builder::new()
+            .prefix("test_customized_migration")
+            .tempdir()
+            .unwrap();
+        let config = DBConfig {
+            path: tmp_dir.as_ref().to_path_buf(),
+            ..Default::default()
+        };
+
+        {
+            let mut migrations = Migrations::default();
+            migrations.add_migration(Box::new(DefaultMigration::new("20191116225943")));
+            let db = migrations.migrate(RocksDB::open(&config, 1)).unwrap();
+            let txn = db.transaction();
+            txn.put(COLUMN, &[1, 1], &[1, 1, 1]).unwrap();
+            txn.put(COLUMN, &[2, 2], &[2, 2, 2]).unwrap();
+            txn.commit().unwrap();
+        }
+        {
+            let mut migrations = Migrations::default();
+            migrations.add_migration(Box::new(DefaultMigration::new("20191116225943")));
+            migrations.add_migration(Box::new(CustomizedMigration));
+            let db = migrations.migrate(RocksDB::open(&config, 1)).unwrap();
+            assert!(
+                vec![1u8, 1, 1, 1].as_slice()
+                    == db.get_pinned(COLUMN, &[1, 1]).unwrap().unwrap().as_ref()
+            );
+            assert!(
+                vec![2u8, 2, 2, 1].as_slice()
+                    == db.get_pinned(COLUMN, &[2, 2]).unwrap().unwrap().as_ref()
+            );
+            assert_eq!(
+                VERSION.as_bytes(),
+                db.get_pinned_defalt(VERSION_KEY)
+                    .unwrap()
+                    .unwrap()
+                    .to_vec()
+                    .as_slice()
+            );
+        }
+    }
+}
```